roles/adoption_osp_deploy/tasks/main.yml

```diff
@@ -0,0 +1,155 @@
+# License for the specific language governing permissions and limitations
+# under the License.
+
+- name: Prepare undercloud enviornment
+  ansible.builtin.import_tasks: prepare_undercloud.yml
+
+- name: Generate configuration files
+  ansible.builtin.import_tasks: config_files.yml
+
+- name: Group undercloud tasks
+  vars:
+    _skip_tags: >-
+      {% if cifmw_adoption_osp_deploy_skip_stages is defined and
+         cifmw_adoption_osp_deploy_skip_stages is string -%}
+      {{ cifmw_adoption_osp_deploy_skip_stages | ansible.builtin.split(",") }}
+      {% elif cifmw_adoption_osp_deploy_skip_stages is defined and
+         cifmw_adoption_osp_deploy_skip_stages is not string and
+         cifmw_adoption_osp_deploy_skip_stages is iterable -%}
+      {{ cifmw_adoption_osp_deploy_skip_stages }}
+      {% else -%}
+      {{ [] }}
+      {% endif -%}
+    _step_name: "undercloud"
+  when:
+    - _step_name not in _skip_tags
+  block:
+    - name: Stop before running pre undercloud hooks
+      vars:
+        _stage_stopper: "before_pre_hook_undercloud"
+      when:
+        - cifmw_adoption_osp_deploy_stopper is defined
+        - cifmw_adoption_osp_deploy_stopper in _stage_stopper
+      ansible.builtin.fail:
+        msg: "Failing on demand {{ cifmw_adoption_osp_deploy_stopper }}"
+
+    - name: Run pre undercloud deploy hooks
+      vars:
+        hooks: "{{ _adoption_source_scenario.pre_uc_run | default([]) }}"
+        step: "pre_undercloud"
+      ansible.builtin.include_role:
+        name: run_hook
+
+    - name: Stop before deploying undercloud
+      vars:
+        _stage_stopper: "before_deploy_undercloud"
+      when:
+        - cifmw_adoption_osp_deploy_stopper is defined
+        - cifmw_adoption_osp_deploy_stopper in _stage_stopper
+      ansible.builtin.fail:
+        msg: "Failing on demand {{ cifmw_adoption_osp_deploy_stopper }}"
+
+    - name: Deploy undercloud
+      ansible.builtin.import_tasks: deploy_undercloud.yml
+
+    - name: Stop after deploying undercloud
+      vars:
+        _stage_stopper: "after_deploy_undercloud"
+      when:
+        - cifmw_adoption_osp_deploy_stopper is defined
+        - cifmw_adoption_osp_deploy_stopper in _stage_stopper
+      ansible.builtin.fail:
+        msg: "Failing on demand {{ cifmw_adoption_osp_deploy_stopper }}"
+
+    - name: Run post undercloud deploy hooks
+      vars:
+        hooks: "{{ _adoption_source_scenario.post_uc_run | default([]) }}"
+        step: "post_undercloud"
+      ansible.builtin.include_role:
+        name: run_hook
+
+    - name: Stop after running post undercloud hooks
+      vars:
+        _stage_stopper: "after_post_hook_undercloud"
+      when:
+        - cifmw_adoption_osp_deploy_stopper is defined
+        - cifmw_adoption_osp_deploy_stopper in _stage_stopper
+      ansible.builtin.fail:
+        msg: "Failing on demand {{ cifmw_adoption_osp_deploy_stopper }}"
+
+- name: Group overcloud tasks
+  vars:
```

Comment:
ditto, `tags: undercloud` would look nicer imho - but beware of zuul context as put in the first comment.